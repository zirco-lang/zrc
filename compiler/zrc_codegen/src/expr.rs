use anyhow::{bail, Context as _};

use super::{
    cg_alloc, cg_load, cg_store, determine_order_of_struct, get_llvm_typename, BasicBlock, CgScope,
    FunctionCg, ModuleCg,
};

/// Returns a register containing a pointer to the place inputted
pub fn cg_place(
    module: &mut ModuleCg,
    cg: &mut FunctionCg,
    bb: &BasicBlock,
    scope: &CgScope,
    place: zrc_typeck::tast::expr::Place,
) -> anyhow::Result<(String, BasicBlock)> {
    // Again, produces a REGISTER holding the pointer.

    use zrc_typeck::tast::expr::PlaceKind;

    Ok(match place.1 {
        PlaceKind::Variable(x) => {
            let reg = scope
                .get(&x)
                .with_context(|| format!("Identifier {} not found in scope", x))?
                .clone();
            (reg, bb.clone())
        }
        PlaceKind::Deref(x) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x)?;
            (x_ptr, bb)
        }
        PlaceKind::Index(x, index) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x.clone())?;
            let (index_ptr, bb) = cg_expr(module, cg, &bb, scope, *index.clone())?;

            let x_typename = get_llvm_typename(x.clone().0);

            let x_reg = cg_load(cg, &bb, &x_typename, &x_ptr)?;

            let index_typename = get_llvm_typename(index.0.clone());

            let index_reg = cg_load(cg, &bb, &index_typename, &index_ptr)?;

            let result_typename = match x.0 {
                zrc_typeck::tast::ty::Type::Ptr(x) => get_llvm_typename(*x),
                _ => unreachable!(), // per typeck, this is always a pointer
            };

            let result_reg = cg.new_reg();

            #[allow(clippy::uninlined_format_args)] // for line length
            bb.add_instruction(
                cg,
                &format!(
                    "{} = getelementptr {}, {} {}, {} {}",
                    result_reg, result_typename, x_typename, x_reg, index_typename, index_reg
                ),
            )?;

            (result_reg, bb)
        }
        PlaceKind::Arrow(x, key) => {
            let (x_ptr, bb) = cg_place(module, cg, bb, scope, *x.clone())?;

            let x_typename = get_llvm_typename(x.0.clone());

            let x_reg = cg_load(cg, &bb, &x_typename, &x_ptr)?;

            let result_typename = match x.0.clone() {
                zrc_typeck::tast::ty::Type::Ptr(x) => get_llvm_typename(*x),
                _ => unreachable!(), // per typeck, this is always a pointer
            };

            let result_reg = cg.new_reg();

            let key_idx = match x.0.clone() {
                zrc_typeck::tast::ty::Type::Struct(entries) => {
                    determine_order_of_struct(entries)
                        .into_iter()
                        .enumerate()
                        .find(|(_, (k, _))| k == &key)
                        .with_context(|| format!("Struct {} has no field {}", x.0, key))?
                        .0
                }
                _ => unreachable!(), // per typeck, this is always a struct
            };

            #[allow(clippy::uninlined_format_args)] // for line length
            bb.add_instruction(
                cg,
                &format!(
                    "{} = getelementptr {}, {} {}, i32 0, i32 0, i32 {}",
                    result_reg, result_typename, x_typename, x_reg, key_idx
                ),
            )?;

            (result_reg, bb)
        }
        PlaceKind::Dot(x, key) => {
            let (x_ptr, bb) = cg_place(module, cg, bb, scope, *x.clone())?;

            let x_typename = get_llvm_typename(x.0.clone());

            let x_reg = cg_load(cg, &bb, &x_typename, &x_ptr)?;

            let result_typename = match x.0.clone() {
                zrc_typeck::tast::ty::Type::Struct(entries) => get_llvm_typename(
                    entries
                        .get(&key)
                        .with_context(|| format!("Struct {} has no field {}", x.0.clone(), key))?
                        .clone(),
                ),
                _ => unreachable!(), // per typeck, this is always a struct
            };

            let result_reg = cg.new_reg();

            let key_idx = match x.0.clone() {
                zrc_typeck::tast::ty::Type::Struct(entries) => {
                    determine_order_of_struct(entries)
                        .into_iter()
                        .enumerate()
                        .find(|(_, (k, _))| k == &key)
                        .with_context(|| format!("Struct {} has no field {}", x.0, key))?
                        .0
                }
                _ => unreachable!(), // per typeck, this is always a struct
            };

            #[allow(clippy::uninlined_format_args)] // for line length
            bb.add_instruction(
                cg,
                &format!(
                    "{} = getelementptr {}, {} {}, i32 0, i32 {}",
                    result_reg, result_typename, x_typename, x_reg, key_idx
                ),
            )?;

            (result_reg, bb)
        }
    })
}

/// Returns a register containing a pointer to the result of the expression
/// along with the basic block to continue adding statements from.
pub fn cg_expr(
    module: &mut ModuleCg,
    cg: &mut FunctionCg,
    bb: &BasicBlock,
    scope: &CgScope,
    expr: zrc_typeck::tast::expr::TypedExpr,
) -> anyhow::Result<(String, BasicBlock)> {
    use zrc_typeck::tast::expr::TypedExprKind;
    // Remember you must ALWAYS return a POINTER register. This is an intentional
    // design choice for now. llvm can optimize away the pointer if it wants to.

    Ok(match expr.1 {
        TypedExprKind::Ternary(condition, lhs, rhs) => {
            let (condition_ptr, bb) = cg_expr(module, cg, bb, scope, *condition)?;
            // condition_ptr: i1*

            let if_true_bb = cg.new_bb();
            let if_false_bb = cg.new_bb();

            // load the condition
            let condition_reg = cg_load(cg, &bb, "i1", &condition_ptr)?;
            // and branch on it
            bb.add_instruction(
                cg,
                &format!("br i1 {condition_reg}, label {if_true_bb}, label {if_false_bb}"),
            )?;

            // generate the expressions on both sides
            let (if_true_ptr, if_true_bb) = cg_expr(module, cg, &if_true_bb, scope, *lhs)?;
            let (if_false_ptr, if_false_bb) = cg_expr(module, cg, &if_false_bb, scope, *rhs)?;
            // if_true_ptr is a pointer accessible on the true side and vice versa

            let terminating_bb = cg.new_bb();

            // branch to the terminating bb
            if_true_bb.add_instruction(cg, &format!("br label {terminating_bb}"))?;
            if_false_bb.add_instruction(cg, &format!("br label {terminating_bb}"))?;

            let result_ptr = cg.new_reg();

            // now generate a phi node
            #[allow(clippy::uninlined_format_args)] // for line length
            terminating_bb.add_instruction(
                cg,
                &format!(
                    "{} = phi ptr [{}, {}], [{}, {}]",
                    result_ptr, if_false_ptr, if_false_bb, if_true_ptr, if_true_bb
                ),
            )?;

            (result_ptr, terminating_bb)
        }

        TypedExprKind::NumberLiteral(n) => {
            let typename = get_llvm_typename(expr.0.clone());
            let reg = cg_alloc(cg, bb, &typename);
            cg_store(cg, bb, &typename, &reg, &n)?;
            (reg, bb.clone())
        }
        TypedExprKind::BooleanLiteral(b) => {
            let typename = get_llvm_typename(expr.0.clone());
            let reg = cg_alloc(cg, bb, &typename);
            cg_store(cg, bb, &typename, &reg, &b.to_string())?;
            (reg, bb.clone())
        }

        TypedExprKind::Comma(lhs, rhs) => {
            let (_, bb) = cg_expr(module, cg, bb, scope, *lhs)?;
            let (rhs_ptr, bb) = cg_expr(module, cg, &bb, scope, *rhs)?;
            (rhs_ptr, bb)
        }

        TypedExprKind::BinaryBitwise(op, lhs, rhs) => {
            let (lhs_ptr, bb) = cg_expr(module, cg, bb, scope, *lhs)?;
            let (rhs_ptr, bb) = cg_expr(module, cg, &bb, scope, *rhs)?;

            let result_typename = get_llvm_typename(expr.0.clone());
            let result_ptr = cg_alloc(cg, &bb, &result_typename);

            let lhs_reg = cg_load(cg, &bb, &result_typename, &lhs_ptr)?;
            let rhs_reg = cg_load(cg, &bb, &result_typename, &rhs_ptr)?;

            let op = match op {
                zrc_typeck::tast::expr::BinaryBitwise::And => "and",
                zrc_typeck::tast::expr::BinaryBitwise::Or => "or",
                zrc_typeck::tast::expr::BinaryBitwise::Xor => "xor",
                zrc_typeck::tast::expr::BinaryBitwise::Shl => "shl",
                zrc_typeck::tast::expr::BinaryBitwise::Shr => "lshr", // should it be ashr?
            };

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = {op} {result_typename} {lhs_reg}, {rhs_reg}"),
            )?;
            cg_store(cg, &bb, &result_typename, &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }

        TypedExprKind::Equality(op, lhs, rhs) => {
            // it's either *T == *U or iN == iN

            let (lhs_ptr, bb) = cg_expr(module, cg, bb, scope, *lhs.clone())?;
            let (rhs_ptr, bb) = cg_expr(module, cg, &bb, scope, *rhs)?;

            let operand_typename = get_llvm_typename(lhs.0.clone());

            let lhs_reg = cg_load(cg, &bb, &operand_typename, &lhs_ptr)?;
            let rhs_reg = cg_load(cg, &bb, &operand_typename, &rhs_ptr)?;

            let op = match op {
                zrc_typeck::tast::expr::Equality::Eq => "icmp eq",
                zrc_typeck::tast::expr::Equality::Neq => "icmp ne",
            };

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = {op} {operand_typename} {lhs_reg}, {rhs_reg}"),
            )?;

            let result_ptr = cg_alloc(cg, &bb, "i1");
            cg_store(cg, &bb, "i1", &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }

        TypedExprKind::Comparison(op, lhs, rhs) => {
            let (lhs_ptr, bb) = cg_expr(module, cg, bb, scope, *lhs.clone())?;
            let (rhs_ptr, bb) = cg_expr(module, cg, &bb, scope, *rhs)?;

            let result_typename = get_llvm_typename(lhs.0.clone());

            let lhs_reg = cg_load(cg, &bb, &result_typename, &lhs_ptr)?;
            let rhs_reg = cg_load(cg, &bb, &result_typename, &rhs_ptr)?;

            // the operands are integers
            let op = match (op, expr.0.clone()) {
                (zrc_typeck::tast::expr::Comparison::Lt, ty) if ty.is_signed_integer() => {
                    "icmp slt"
                }
                (zrc_typeck::tast::expr::Comparison::Gt, ty) if ty.is_signed_integer() => {
                    "icmp sgt"
                }
                (zrc_typeck::tast::expr::Comparison::Lte, ty) if ty.is_signed_integer() => {
                    "icmp sle"
                }
                (zrc_typeck::tast::expr::Comparison::Gte, ty) if ty.is_signed_integer() => {
                    "icmp sge"
                }

                (zrc_typeck::tast::expr::Comparison::Lt, _) => "icmp ult",
                (zrc_typeck::tast::expr::Comparison::Gt, _) => "icmp ugt",
                (zrc_typeck::tast::expr::Comparison::Lte, _) => "icmp ule",
                (zrc_typeck::tast::expr::Comparison::Gte, _) => "icmp uge",
            };

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = {op} {result_typename} {lhs_reg}, {rhs_reg}"),
            )?;

            let result_ptr = cg_alloc(cg, &bb, "i1");
            cg_store(cg, &bb, "i1", &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }
        TypedExprKind::Arithmetic(op, lhs, rhs) => {
            let (lhs_ptr, bb) = cg_expr(module, cg, bb, scope, *lhs)?;
            let (rhs_ptr, bb) = cg_expr(module, cg, &bb, scope, *rhs)?;

            let result_typename = get_llvm_typename(expr.0.clone());

            let lhs_reg = cg_load(cg, &bb, &result_typename, &lhs_ptr)?;
            let rhs_reg = cg_load(cg, &bb, &result_typename, &rhs_ptr)?;

            let op = match op {
                zrc_typeck::tast::expr::Arithmetic::Addition => "add",
                zrc_typeck::tast::expr::Arithmetic::Subtraction => "sub",
                zrc_typeck::tast::expr::Arithmetic::Multiplication => "mul",
                zrc_typeck::tast::expr::Arithmetic::Division => "sdiv",
                zrc_typeck::tast::expr::Arithmetic::Modulo => "srem",
            };

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = {op} {result_typename} {lhs_reg}, {rhs_reg}"),
            )?;

            let result_ptr = cg_alloc(cg, &bb, &result_typename);
            cg_store(cg, &bb, &result_typename, &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }
        TypedExprKind::Logical(op, lhs, rhs) => {
            let (lhs_ptr, bb) = cg_expr(module, cg, bb, scope, *lhs)?;
            let (rhs_ptr, bb) = cg_expr(module, cg, &bb, scope, *rhs)?;

            let result_typename = get_llvm_typename(expr.0.clone());

            let lhs_reg = cg_load(cg, &bb, &result_typename, &lhs_ptr)?;
            let rhs_reg = cg_load(cg, &bb, &result_typename, &rhs_ptr)?;

            // "and/or i1" works for bools
            let op = match op {
                zrc_typeck::tast::expr::Logical::And => "and",
                zrc_typeck::tast::expr::Logical::Or => "or",
            };

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = {op} {result_typename} {lhs_reg}, {rhs_reg}"),
            )?;

            let result_ptr = cg_alloc(cg, &bb, &result_typename);
            cg_store(cg, &bb, &result_typename, &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }

        TypedExprKind::UnaryBitwiseNot(x) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x)?;

            let result_typename = get_llvm_typename(expr.0.clone());

            let x_reg = cg_load(cg, &bb, &result_typename, &x_ptr)?;

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = xor {result_typename} {x_reg}, -1"),
            )?;

            let result_ptr = cg_alloc(cg, &bb, &result_typename);
            cg_store(cg, &bb, &result_typename, &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }

        TypedExprKind::UnaryNot(x) => {
            // x is bool
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x)?;

            let result_typename = get_llvm_typename(expr.0.clone());

            let x_reg = cg_load(cg, &bb, &result_typename, &x_ptr)?;

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = xor {result_typename} {x_reg}, 1"),
            )?;

            let result_ptr = cg_alloc(cg, &bb, &result_typename);

            cg_store(cg, &bb, &result_typename, &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }

        TypedExprKind::UnaryMinus(x) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x)?;

            let result_typename = get_llvm_typename(expr.0.clone());

            let x_reg = cg_load(cg, &bb, &result_typename, &x_ptr)?;

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = sub {result_typename} 0, {x_reg}"),
            )?;

            let result_ptr = cg_alloc(cg, &bb, &result_typename);

            cg_store(cg, &bb, &result_typename, &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }

        TypedExprKind::UnaryAddressOf(x) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x)?;

            let result_typename = get_llvm_typename(expr.0.clone());

            let result_ptr = cg_alloc(cg, &bb, &result_typename);

            cg_store(cg, &bb, &result_typename, &result_ptr, &x_ptr)?;

            (result_ptr, bb)
        }

        TypedExprKind::UnaryDereference(x) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x)?;

            let result_typename = get_llvm_typename(expr.0.clone());

            let x_reg = cg_load(cg, &bb, &result_typename, &x_ptr)?;

            let result_ptr = cg_alloc(cg, &bb, &result_typename);

            cg_store(cg, &bb, &result_typename, &result_ptr, &x_reg)?;

            (result_ptr, bb)
        }

        TypedExprKind::Assignment(place, value) => {
            let (place_ptr, bb) = cg_place(module, cg, bb, scope, *place)?;
            let (value_ptr, bb) = cg_expr(module, cg, &bb, scope, *value.clone())?;

            let value_typename = get_llvm_typename(value.0.clone());

            let value_reg = cg_load(cg, &bb, &value_typename, &value_ptr)?;

            cg_store(cg, &bb, &value_typename, &place_ptr, &value_reg)?;

            (place_ptr, bb)
        }

        TypedExprKind::Identifier(id) => {
            let reg = scope
                .get(&id)
                .with_context(|| format!("Identifier {} not found in scope", id))?
                .clone();
            (reg, bb.clone())
        }

        TypedExprKind::Index(x, index) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x.clone())?;
            let (index_ptr, bb) = cg_expr(module, cg, &bb, scope, *index.clone())?;

            let x_typename = get_llvm_typename(x.clone().0);

            let x_reg = cg_load(cg, &bb, &x_typename, &x_ptr)?;

            let index_typename = get_llvm_typename(index.0.clone());

            let index_reg = cg_load(cg, &bb, &index_typename, &index_ptr)?;

            let result_typename = match x.0 {
                zrc_typeck::tast::ty::Type::Ptr(x) => get_llvm_typename(*x),
                _ => unreachable!(), // per typeck, this is always a pointer
            };

            let result_reg = cg.new_reg();

            #[allow(clippy::uninlined_format_args)] // for line length
            bb.add_instruction(
                cg,
                &format!(
                    "{} = getelementptr {}, {} {}, {} {}",
                    result_reg, result_typename, x_typename, x_reg, index_typename, index_reg
                ),
            )?;

            (result_reg, bb)
        }

        TypedExprKind::Dot(x, prop) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x.clone())?;

            let x_typename = get_llvm_typename(x.0.clone());

            let x_reg = cg_load(cg, &bb, &x_typename, &x_ptr)?;

            let result_typename = match x.0.clone() {
                zrc_typeck::tast::ty::Type::Struct(entries) => get_llvm_typename(
                    entries
                        .get(&prop)
                        .with_context(|| format!("Struct {} has no field {}", x.0.clone(), prop))?
                        .clone(),
                ),
                _ => unreachable!(), // per typeck, this is always a struct
            };

            let result_reg = cg.new_reg();

            let key_idx = match x.0.clone() {
                zrc_typeck::tast::ty::Type::Struct(entries) => {
                    determine_order_of_struct(entries)
                        .into_iter()
                        .enumerate()
                        .find(|(_, (k, _))| k == &prop)
                        .with_context(|| format!("Struct {} has no field {}", x.0, prop))?
                        .0
                }
                _ => unreachable!(), // per typeck, this is always a struct
            };

            #[allow(clippy::uninlined_format_args)] // for line length
            bb.add_instruction(
                cg,
                &format!(
                    "{} = getelementptr {}, {} {}, i32 0, i32 {}",
                    result_reg, result_typename, x_typename, x_reg, key_idx
                ),
            )?;

            (result_reg, bb)
        }

        TypedExprKind::Arrow(x, prop) => {
            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x.clone())?;

            let x_typename = get_llvm_typename(x.0.clone());

            let x_reg = cg_load(cg, &bb, &x_typename, &x_ptr)?;

            let result_typename = match x.0.clone() {
                zrc_typeck::tast::ty::Type::Ptr(x) => get_llvm_typename(*x),
                _ => unreachable!(), // per typeck, this is always a pointer
            };

            let result_reg = cg.new_reg();

            let key_idx = match x.0.clone() {
                zrc_typeck::tast::ty::Type::Struct(entries) => {
                    determine_order_of_struct(entries)
                        .into_iter()
                        .enumerate()
                        .find(|(_, (k, _))| k == &prop)
                        .with_context(|| format!("Struct {} has no field {}", x.0.clone(), prop))?
                        .0
                }
                _ => unreachable!(), // per typeck, this is always a struct
            };

            #[allow(clippy::uninlined_format_args)] // for line length
            bb.add_instruction(
                cg,
                &format!(
                    "{} = getelementptr {}, {} {}, i32 0, i32 0, i32 {}",
                    result_reg, result_typename, x_typename, x_reg, key_idx
                ),
            )?;

            (result_reg, bb)
        }

        TypedExprKind::Call(f, args) => {
            let (f_ptr, bb) = cg_expr(module, cg, bb, scope, *f)?;

            let args = args
                .into_iter()
                .map(|arg| {
                    let (arg_ptr, bb) = cg_expr(module, cg, &bb, scope, arg.clone())?;
                    let arg_typename = get_llvm_typename(arg.0);
                    let arg_reg = cg_load(cg, &bb, &arg_typename, &arg_ptr)?;
                    Ok(format!("{arg_typename} {arg_reg}"))
                })
                .collect::<anyhow::Result<Vec<_>>>()?;

            let args = args.join(", ");

            let result_typename = get_llvm_typename(expr.0.clone());

            if result_typename == "void" {
                bb.add_instruction(cg, &format!("call {result_typename} {f_ptr}({args})"))?;

                let undef = cg_alloc(cg, &bb, "i8");

                // store the undefined value here as `void` results should never be loaded but
                // our code generator expects a loadable value
                cg_store(cg, &bb, "i8", &undef, "undef")?;

                (undef, bb)
            } else {
                let result_reg = cg.new_reg();

                bb.add_instruction(
                    cg,
                    &format!("{result_reg} = call {result_typename} {f_ptr}({args})"),
                )?;

                let result_ptr = cg_alloc(cg, &bb, &result_typename);

                cg_store(cg, &bb, &result_typename, &result_ptr, &result_reg)?;

                (result_ptr, bb)
            }
        }

        TypedExprKind::Cast(x, t) => {
            use zrc_typeck::tast::ty::Type::*;

            // The legendary Cast Table. Contains the cast opcode used for T -> U
            let cast_opcode = match (x.0.clone(), t.clone()) {
                (x, y) if x == y => "bitcast",
                // signed -> signed = sext
                // signed -> unsigned = sext
                // unsigned -> signed = zext
                // unsigned -> unsigned = zext
                // ptr -> ptr = bitcast
                // T -> T = bitcast
                // int -> ptr = inttoptr
                // ptr -> int = ptrtoint
                // int -> fn = inttoptr
                // fn -> int = ptrtoint
                (Bool, I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64) => "zext",
                (I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64, Bool) => "trunc",

                (I8, I16 | U16 | I32 | U32 | I64 | U64) => "sext",
                (U8, I16 | U16 | I32 | U32 | I64 | U64) => "zext",
                (I16, I32 | U32 | I64 | U64) => "sext",
                (U16, I32 | U32 | I64 | U64) => "zext",
                (I32, I64 | U64) => "sext",
                (U32, I64 | U64) => "zext",

                (I16 | U16 | I32 | U32 | I64 | U64, I8) => "trunc",
                (I16 | U16 | I32 | U32 | I64 | U64, U8) => "trunc",
                (I32 | U32 | I64 | U64, I16) => "trunc",
                (I32 | U32 | I64 | U64, U16) => "trunc",
                (I64 | U64, I32) => "trunc",
                (I64 | U64, U32) => "trunc",

                // also handle things like I32 => U32
                // is this even the correct way to do it?
                (I8, U8) | (U8, I8) => "bitcast",
                (I16, U16) | (U16, I16) => "bitcast",
                (I32, U32) | (U32, I32) => "bitcast",
                (I64, U64) | (U64, I64) => "bitcast",

                (Bool, x) if x.is_signed_integer() => "sext",
                (Bool, x) if x.is_unsigned_integer() => "zext",

                (Ptr(_), Ptr(_)) => "bitcast",
                (Ptr(_), I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64) => "ptrtoint",
                (I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64, Ptr(_)) => "inttoptr",
                (I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64, Fn(_, _)) => "inttoptr",
                (Fn(_, _), I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64) => "ptrtoint",
                _ => bail!("invalid cast from {} to {}", x.0, t),
            };

            let (x_ptr, bb) = cg_expr(module, cg, bb, scope, *x.clone())?;

            let x_typename = get_llvm_typename(x.0.clone());

            let x_reg = cg_load(cg, &bb, &x_typename, &x_ptr)?;

            let result_typename = get_llvm_typename(expr.0.clone());

            let result_reg = cg.new_reg();

            bb.add_instruction(
                cg,
                &format!("{result_reg} = {cast_opcode} {x_typename} {x_reg} to {result_typename}"),
            )?;

            let result_ptr = cg_alloc(cg, &bb, &result_typename);

            cg_store(cg, &bb, &result_typename, &result_ptr, &result_reg)?;

            (result_ptr, bb)
        }

        TypedExprKind::StringLiteral(s) => {
            let id = module.global_constant_id.next();
            module.declarations.push(format!(
                "@.str{id} = private constant [{} x i8] c\"{}\\00\"",
                s.len() - 1, // subtract both quotes, add the null
                &s[1..s.len() - 1]
            ));

            let result_reg = cg_alloc(cg, bb, "ptr"); // u8*

            cg_store(cg, bb, "ptr", &result_reg, &format!("@.str{id}"))?;

            (result_reg, bb.clone())
        }
    })
}
