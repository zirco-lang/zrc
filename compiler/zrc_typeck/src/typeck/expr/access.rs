//! type checking for the indexing and access operators

use zrc_diagnostics::{Diagnostic, DiagnosticKind};
use zrc_parser::ast::expr::{Expr, ExprKind};
use zrc_utils::span::{Span, Spannable, Spanned};

use super::{
    super::scope::Scope,
    helpers::{expect, expr_to_place},
    type_expr,
};
use crate::tast::{
    expr::{TypedExpr, TypedExprKind},
    ty::Type as TastType,
};

/// Typeck an index expr
pub fn type_expr_index<'input>(
    scope: &Scope<'input, '_>,
    expr_span: Span,
    ptr: Expr<'input>,
    offset: Expr<'input>,
) -> Result<TypedExpr<'input>, Diagnostic> {
    let ptr_t = type_expr(scope, ptr)?;
    let offset_t = type_expr(scope, offset)?;

    expect(
        offset_t.inferred_type == TastType::Usize,
        "usize".to_string(),
        offset_t.inferred_type.to_string(),
        offset_t.kind.span(),
    )?;

    if let TastType::Ptr(points_to_ty) = ptr_t.inferred_type.clone() {
        Ok(TypedExpr {
            inferred_type: *points_to_ty,
            kind: TypedExprKind::Index(Box::new(ptr_t), Box::new(offset_t)).in_span_no_file(expr_span),
        })
    } else {
        Err(
            DiagnosticKind::CannotIndexIntoNonPointer(ptr_t.inferred_type.to_string())
                .error_in(expr_span),
        )
    }
}

/// Typeck a dot expr
pub fn type_expr_dot<'input>(
    scope: &Scope<'input, '_>,
    expr_span: Span,
    obj: Expr<'input>,
    key: Spanned<&'input str>,
) -> Result<TypedExpr<'input>, Diagnostic> {
    let obj_span = obj.0.span();
    let obj_t = type_expr(scope, obj)?;

    if let TastType::Struct(fields) | TastType::Union(fields) = obj_t.inferred_type.clone() {
        if let Some(ty) = fields.get(key.value()) {
            Ok(TypedExpr {
                inferred_type: ty.clone(),
                kind: TypedExprKind::Dot(Box::new(expr_to_place(obj_span, obj_t)?), key)
                    .in_span_no_file(expr_span),
            })
        } else {
            Err(DiagnosticKind::StructOrUnionDoesNotHaveMember(
                obj_t.inferred_type.to_string(),
                key.into_value().to_string(),
            )
            .error_in(expr_span))
        }
    } else {
        Err(
            DiagnosticKind::StructMemberAccessOnNonStruct(obj_t.inferred_type.to_string())
                .error_in(expr_span),
        )
    }
}

/// Typeck an arrow expr
pub fn type_expr_arrow<'input>(
    scope: &Scope<'input, '_>,
    expr_span: Span,
    obj: Box<Expr<'input>>,
    key: Spanned<&'input str>,
) -> Result<TypedExpr<'input>, Diagnostic> {
    let obj_t = type_expr(scope, *obj.clone())?;

    if let TastType::Ptr(_) = obj_t.inferred_type {
        type_expr(
            scope,
            Expr(Spanned::from_span(
                Span::from_positions(obj.0.start(), key.end()),
                ExprKind::Dot(
                    Box::new(Expr(
                        obj.0.span().containing_no_file(ExprKind::UnaryDereference(obj)),
                    )),
                    key,
                ),
            )),
        )
    } else {
        Err(
            DiagnosticKind::CannotDereferenceNonPointer(obj_t.inferred_type.to_string())
                .error_in(expr_span),
        )
    }
}
