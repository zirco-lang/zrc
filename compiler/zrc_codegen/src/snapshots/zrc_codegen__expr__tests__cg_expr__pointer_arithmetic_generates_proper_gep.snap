---
source: compiler/zrc_codegen/src/expr.rs
description: "fn test() {\n    let x: *i32;\n\n    // TEST: should create a GEP that is the same as \"x[4]\"\n    let y = x + 4 as usize;\n    // TEST: and the same, with -4:\n    let z = x - 4 as usize;\n\n    return;\n}\n"
expression: resulting_ir
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

define void @test() !dbg !3 {
entry:
  %let_z = alloca ptr, align 8
  %let_y = alloca ptr, align 8
  %let_x = alloca ptr, align 8
  call void @llvm.dbg.declare(metadata ptr %let_x, metadata !8, metadata !DIExpression()), !dbg !15
  call void @llvm.dbg.declare(metadata ptr %let_y, metadata !13, metadata !DIExpression()), !dbg !16
  %load = load ptr, ptr %let_x, align 8, !dbg !17
  %ptr_add = getelementptr i32, ptr %load, i64 4, !dbg !18
  store ptr %ptr_add, ptr %let_y, align 8, !dbg !16
  call void @llvm.dbg.declare(metadata ptr %let_z, metadata !14, metadata !DIExpression()), !dbg !19
  %load1 = load ptr, ptr %let_x, align 8, !dbg !20
  %ptr_sub = getelementptr i32, ptr %load1, i64 -4, !dbg !21
  store ptr %ptr_sub, ptr %let_z, align 8, !dbg !19
  ret void, !dbg !22
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare void @llvm.dbg.declare(metadata, metadata, metadata) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 1, type: !4, scopeLine: 1, spFlags: DISPFlagDefinition, unit: !1, retainedNodes: !7)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DIBasicType(name: "void")
!7 = !{!8, !13, !14}
!8 = !DILocalVariable(name: "x", scope: !9, file: !2, line: 2, type: !11)
!9 = distinct !DILexicalBlock(scope: !10, file: !2, line: 1, column: 11)
!10 = distinct !DILexicalBlock(scope: !3, file: !2, line: 1, column: 11)
!11 = !DIDerivedType(tag: DW_TAG_pointer_type, name: "\22i32\22", baseType: !12, dwarfAddressSpace: 0)
!12 = !DIBasicType(name: "i32")
!13 = !DILocalVariable(name: "y", scope: !9, file: !2, line: 5, type: !11)
!14 = !DILocalVariable(name: "z", scope: !9, file: !2, line: 7, type: !11)
!15 = !DILocation(line: 2, column: 9, scope: !9)
!16 = !DILocation(line: 5, column: 9, scope: !9)
!17 = !DILocation(line: 5, column: 13, scope: !9)
!18 = !DILocation(line: 5, column: 17, scope: !9)
!19 = !DILocation(line: 7, column: 9, scope: !9)
!20 = !DILocation(line: 7, column: 13, scope: !9)
!21 = !DILocation(line: 7, column: 17, scope: !9)
!22 = !DILocation(line: 9, column: 5, scope: !9)

